src/shmem_team.c

```diff
@@ -0,0 +1,460 @@
+/* -*- C -*-
+ *
+ * Copyright (c) 2019 Intel Corporation. All rights reserved.
+ * This software is available to you under the BSD license.
+ *
+ * This file is part of the Sandia OpenSHMEM software package. For license
+ * information, see the LICENSE file in the top level directory of the
+ * distribution.
+ *
+ */
+
+
+#include "config.h"
+
+#include "shmemx.h"
+#include "shmem_team.h"
+#include "shmem_collectives.h"
+#include "shmem_remote_pointer.h"
+
+#include <math.h>
+
+#define SHMEMX_TEAM_WORLD_INDEX   0
+#define SHMEMX_TEAM_SHARED_INDEX  1
+
+shmem_internal_team_t shmem_internal_team_world;
+shmemx_team_t SHMEMX_TEAM_WORLD = (shmemx_team_t) &shmem_internal_team_world;
+
+shmem_internal_team_t shmem_internal_team_shared;
+shmemx_team_t SHMEMX_TEAM_SHARED = (shmemx_team_t) &shmem_internal_team_shared;
+
+shmem_internal_team_t **shmem_internal_team_pool;
+long *shmem_internal_psync_pool;
+long *shmem_internal_psync_barrier_pool;
+static unsigned char *psync_pool_avail;
+static unsigned char *psync_pool_avail_reduced;
+
+shmem_internal_team_op_t shmem_internal_team_sync_type     = SYNC;
+shmem_internal_team_op_t shmem_internal_team_bcast_type    = BCAST;
+shmem_internal_team_op_t shmem_internal_team_reduce_type   = REDUCE;
+shmem_internal_team_op_t shmem_internal_team_collect_type  = COLLECT;
+shmem_internal_team_op_t shmem_internal_team_alltoall_type = ALLTOALL;
+
+static inline
+int check_stride(int pe, int *start, int *stride, int *size)
```

Comment:
This function /really/ needs a top level comment explaining what it does.  Perhaps a more detailed name, as well, since it does more than a simple "check".  :)

Reply:
Maybe `check_add_pe_to_shared_team`?

Reply:
Ha!  It definitely needs a comment.

How about `check_for_linear_stride` since it probably works for any PE subset?

Reply:
And does this make sense?

```
/* Checks whether a PE has a consistent stride given (start, stride, size).
 * This function is useful within a loop across PE IDs, and sets 'start',
 * 'stride' and 'size' accordingly upon exiting the loop. It also assumes
 * 'start' and 'stride' are initialized to a negative number and 'size' to 0.
 * If an inconsistent stride is found, returns -1. */
```

Reply:
Looks good!